### PR TITLE
Improving Oracle-XE support

### PR DESCRIPTION
**Pull request by justinwyer**
Current challenges:

 - The module only connects using SID and not service name, connecting to PDBs in Oracle 12+ typically requires to use a service name.
 - Username and password cannot be specified.

Changes:

 - Detect SID `:` or service name `/` in the supplied TC JDBC URL and honour this in the underlying JDBC URL.
 - Parse username and password from the supplied TC JDBC URL and pass to the underlying JDBC URL.
 - Switched affected regex patterns to use named groups
 - Enable tests
 - Add tests for both SID and service name connections

Additional notes:

Currently Oracle only provides docker scripts to build your own version of 18.4.0 which has a very slow start time (~7 minutes on a decent laptop) in order to make this useful for test cycles it is typical to start this container and then run a `docker commit` on it. This reduces start time to around 10 seconds. _**The tests now make use of such a container, we should probably retag this container and put it somewhere else other than my docker hub account.**_

### CODE REVIEW

**Comment by justinwyer** [comment]
If you're happy with the PR in principal then I will be happy to update the JDBC documentation to include information on using Oracle XE.

**Comment by kiview** [comment]
Hi @justinwyer,

thanks for raising the PR, I think we wrote yesterday in the YT stream? But you probably misunderstood me, what I meant is, that we already have an open Oracle XE PR, that we would like to merge.

Could you maybe review the already open PR at https://github.com/testcontainers/testcontainers-java/pull/4298 and check if it missing anything, that your PR considers?
We especially wanted to use the `gvenzl/oracle-xe` image, since it is semi-official.

**Comment by kiview** [comment]
Greats, let's proceed like this 👍

**Comment by kiview** [comment]
@justinwyer We have just merged #4298, so feel free to expand on this, if you think something is lacking 🙂

**Comment by justinwyer** [comment]
I've rebased, just have a question on the reasons for disallowing system / sys access?

File: modules/oracle-xe/src/main/java/org/testcontainers/containers/OracleContainer.java [hunk 1]
```diff
@@ -27,6 +28,9 @@
 
     private String username = "test";
     private String password = "test";
+    private String serviceName = DEFAULT_SERVICE_NAME;
+    private String sidOrServiceName = "/";
```
[the remark this answers is not in the document]
Comment (justinwyer) [reply modules/oracle-xe/src/main/java/org/testcontainers/containers/OracleContainer.java:32]: I renamed it to database name as SID and service name are just the `:` or `/`

File: modules/oracle-xe/src/main/java/org/testcontainers/containers/OracleContainer.java [hunk 2]
```diff
@@ -102,14 +103,23 @@ public OracleContainer withPassword(String password) {
         return self();
     }
 
+    @Override
+    public OracleContainer withDatabaseName(String sid) {
```
[the remark this answers is not in the document]
Comment (justinwyer) [reply modules/oracle-xe/src/main/java/org/testcontainers/containers/OracleContainer.java:107]: I renamed it to database name as SID and service name are just the `:` or `/`

File: modules/oracle-xe/src/main/java/org/testcontainers/containers/OracleContainer.java [hunk 3]
```diff
@@ -68,7 +72,7 @@ public String getDriverClassName() {
 
     @Override
     public String getJdbcUrl() {
-        return "jdbc:oracle:thin:" + getUsername() + "/" + getPassword() + "@" + getHost() + ":" + getOraclePort() + "/xepdb1";
+        return "jdbc:oracle:thin:" + getUsername() + "/" + getPassword() + "@" + getHost() + ":" + getOraclePort() + sidOrServiceName + getServiceName();
```
[the remark this answers is not in the document]
Comment (justinwyer) [reply modules/oracle-xe/src/main/java/org/testcontainers/containers/OracleContainer.java]: I don't have the time to pick this up.

**Comment by rnorth** [comment]
Re @justinwyer
> quick thought, should we allow the full image name and tag as part of the JDBC url rather than just the image tag, given the nature of the Oracle image, there would no longer be a need for the testcontainers.properties at all.

This is something that we have discussed internally in the past but deferred. Since a docker image name can be a full host+port+path style string, fitting this into the JDBC URL could become messy unless we do it right.

I'm not saying we'll never do this, just that we've considered it and found it unpleasant enough not to want to rush into doing it. I'd suggest not worrying about it for this PR 😉

File: modules/oracle-xe/src/main/java/org/testcontainers/containers/OracleContainer.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (rnorth) [reply modules/oracle-xe/src/main/java/org/testcontainers/containers/OracleContainer.java]: Noting that this thread is resolved in #4402

**Comment by kiview** [comment]
Hey @justinwyer, I am sorry for letting you and the PR waiting.
With regards to #4402, can we limit the changes in this PR to those in `ConnectionUrl` and `OracleContainerProvider`?

The SID and serviceName topic is a bit more complex I think and is better tackled in a future seperate PR and after this PR and #4402 are merged.

**Comment by rnorth** [comment]
I've taken the liberty of preparing a branch with the URL parsing extracted. I've split out the main change that deals with JDBC URL parsing, and added a bit of further tidying and a minor bugfix. I hope this is OK with you @justinwyer - I'm trying to ensure that you retain credit on the commits/PR.

**Comment by justinwyer** [comment]
Thanks guys, I will close this :)

**Comment by kiview** [comment]
Thanks a lot for your work and bringing this on track @justinwyer!

**Pull request closed without merging**